From 51d0c84e88c60af540ca91c0870a225b045e899c Mon Sep 17 00:00:00 2001
From: Clement Faure <clement.faure@nxp.com>
Date: Thu, 24 May 2018 14:52:30 +0200
Subject: [PATCH 3900/5242] MLK-18025: ARM: imx: Fix suspend initialization
 for Optee on imx7ulp

commit  664161d10af3e6e427b1205a84bc0a00ad9cb26a from
https://source.codeaurora.org/external/imx/linux-imx.git

Before the kernel starts, optee uses M4 SRAM to allocate its
suspend function. When imx7ulp_pm_map_io() executes, the psci
driver and psci_ops.cpu_suspend are not initialized yet. This
causes the memset to always wipe the optee suspend code in the
M4 SRAM.

Signed-off-by: Clement Faure <clement.faure@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/mach-imx/pm-imx7ulp.c |    9 +++------
 1 file changed, 3 insertions(+), 6 deletions(-)

diff --git a/arch/arm/mach-imx/pm-imx7ulp.c b/arch/arm/mach-imx/pm-imx7ulp.c
index 777d3ca..41c7d39 100644
--- a/arch/arm/mach-imx/pm-imx7ulp.c
+++ b/arch/arm/mach-imx/pm-imx7ulp.c
@@ -574,9 +574,6 @@ static int __init imx7ulp_dt_find_lpsram(unsigned long node, const char *uname,
 
 void __init imx7ulp_pm_map_io(void)
 {
-	if (psci_ops.cpu_suspend) {
-		return;
-	}
 	/*
 	 * Get the address of IRAM or OCRAM to be used by the low
 	 * power code from the device tree.
@@ -588,9 +585,6 @@ void __init imx7ulp_pm_map_io(void)
 		pr_warn("No valid ocram available for suspend/resume!\n");
 		return;
 	}
-
-	/* Set all entries to 0 except first 3 words reserved for M4. */
-	memset((void *)iram_tlb_base_addr, 0, MX7ULP_IRAM_TLB_SIZE);
 }
 
 void __init imx7ulp_pm_common_init(const struct imx7ulp_pm_socdata
@@ -610,6 +604,9 @@ void __init imx7ulp_pm_common_init(const struct imx7ulp_pm_socdata
 		aips4_base = ioremap(MX7ULP_AIPS4_BASE_ADDR, SZ_1M);
 		aips5_base = ioremap(MX7ULP_AIPS5_BASE_ADDR, SZ_1M);
 	} else {
+		/* Set all entries to 0 except first 3 words reserved for M4. */
+		memset((void *)iram_tlb_base_addr, 0, MX7ULP_IRAM_TLB_SIZE);
+
 		/*
 		 * Make sure the IRAM virtual address has a mapping in the IRAM
 		 * page table.
-- 
1.7.9.5

